pkg/bootstrap/render.go

```diff
@@ -329,6 +331,32 @@ func filesToObjects(files []string, config interface{}) ([]runtime.Object, error
 	return objects, nil
 }
 
+// getKlusterletNamespace returns agentInstallNamespace from config if it is set, otherwise return the
+// input namespace
+func getKlusterletNamespace(namespace string, config *klusterletconfigv1alpha1.KlusterletConfig) string {
+	if config == nil {
+		return namespace
+	}
+	if len(config.Spec.AgentInstallNamespace) == 0 {
+		return namespace
+	}
+	return config.Spec.AgentInstallNamespace
+}
+
+// getKlusterletName returns klusterlet by default, and klusterlet-{cluster name} in hosted mode,
+// and klusterlet-{postfix of install namespace in config}.
+func getKlusterletName(config *klusterletconfigv1alpha1.KlusterletConfig, clusterName, namespace string, mode operatorv1.InstallMode) string {
+	if mode == operatorv1.InstallModeHosted || mode == operatorv1.InstallModeSingletonHosted {
+		return fmt.Sprintf("klusterlet-%s", clusterName)
+	}
+	klusterletNamespace := getKlusterletNamespace(namespace, config)
+	if klusterletNamespace == constants.DefaultKlusterletNamespace {
+		return "klusterlet"
+	}
+	postFix := strings.TrimPrefix(klusterletNamespace, "open-cluster-management-")
+	return fmt.Sprintf("klusterlet-%s", postFix)
```

Comment:
so the klusterlet name is determined by klusterlet namespace?
if the klusterlet namespace changes, we will delete the old klusterlet and create a new one?

Reply:
the klustelet manifestwork delete option is orphan, so if the klusterlet namespace changes, there will be 2 klusterlet CRs on the managed cluster, right? I do not think that is what we expected.